File: samples/dotnet/10-Chaining-Functions/plugins/OrchestratorPlugin/Orchestrator.cs
// Copyright (c) Microsoft. All rights reserved.

using System.ComponentModel;
using Microsoft.SemanticKernel;
using Microsoft.SemanticKernel.Orchestration;
using Newtonsoft.Json.Linq;

namespace Plugins.OrchestratorPlugin;

public class Orchestrator
{
    private readonly IKernel _kernel;

    public Orchestrator(IKernel kernel)
    {
        _kernel = kernel;
    }

    [SKFunction, Description("Routes the request to the appropriate function.")]
    public async Task<string> RouteRequestAsync(
        [Description("The user request")] string input
    )
    {
        // Save the original user request
        string request = input;

        // Retrieve the intent from the user request
        var getIntent = _kernel.Functions.GetFunction("OrchestratorPlugin", "getIntent");
        var getIntentVariables = new ContextVariables
        {
            ["input"] = input,
            ["options"] = "Sqrt, Multiply"
        };
        string intent = (await _kernel.RunAsync(getIntentVariables, getIntent)).GetValue<string>()!.Trim();

        // Call the appropriate function
        ISKFunction MathFunction;
        switch (intent)
        {
            case "Sqrt":
                MathFunction = this._kernel.Functions.GetFunction("MathPlugin", "Sqrt");
                break;
            case "Multiply":
                MathFunction = this._kernel.Functions.GetFunction("MathPlugin", "Multiply");
                break;
            default:
                return "I'm sorry, I don't understand.";
        }

        // Get remaining functions
        var createResponse = this._kernel.Functions.GetFunction("OrchestratorPlugin", "CreateResponse");
        var getNumbers = this._kernel.Functions.GetFunction("OrchestratorPlugin", "GetNumbers");
        var extractNumbersFromJson = this._kernel.Functions.GetFunction("OrchestratorPlugin", "ExtractNumbersFromJson");

        // Run the pipeline
        var output = await this._kernel.RunAsync(
            request,
            getNumbers,
            extractNumbersFromJson,
            MathFunction,
            createResponse
        );

        // Create a new context with the original request
        var pipelineContext = new ContextVariables(request);
        pipelineContext["original_request"] = request;

        // Run the pipeline with create response
        output = await this._kernel.RunAsync(
            pipelineContext,
            getNumbers,
            extractNumbersFromJson,
            MathFunction,
            createResponse
        );

        return output.GetValue<string>()!;
    }

    [SKFunction, Description("Extracts numbers from JSON")]
    public static SKContext ExtractNumbersFromJson(SKContext context)
    {
        JObject numbers = JObject.Parse(context.Variables["input"]);

        // Loop through numbers and add them to the context
        foreach (var number in numbers)
        {
            if (number.Key == "number1")
            {
                context.Variables["input"] = number.Value!.ToString();
                continue;
            }

            context.Variables[number.Key] = number.Value!.ToString();
        }
        return context;
    }
}

File: samples/python/10-Chaining-Functions/plugins/OrchestratorPlugin/Orchestrator.py
import json
from semantic_kernel import ContextVariables, Kernel
from semantic_kernel.skill_definition import (
    sk_function,
)
from semantic_kernel.orchestration.sk_context import SKContext


class Orchestrator:
    def __init__(self, kernel: Kernel):
        self._kernel = kernel

    @sk_function(
        description="Routes the request to the appropriate function",
        name="route_request",
    )
    async def RouteRequest(self, context: SKContext) -> str:
        # Save the original user request
        request = context["input"]

        # Add the list of available functions to the context variables
        variables = ContextVariables()
        variables["input"] = request
        variables["options"] = "Sqrt, Multiply"

        # Retrieve the intent from the user request
        get_intent = self._kernel.skills.get_function("OrchestratorPlugin", "getIntent")
        intent = (
            await self._kernel.run_async(get_intent, input_vars=variables)
        ).result.strip()

        # Prepare the functions to be called in the pipeline
        get_numbers = self._kernel.skills.get_function(
            "OrchestratorPlugin", "GetNumbers"
        )
        extract_numbers_from_json = self._kernel.skills.get_function(
            "OrchestratorPlugin", "ExtractNumbersFromJson"
        )
        create_response = self._kernel.skills.get_function(
            "OrchestratorPlugin", "CreateResponse"
        )

        # Retrieve the correct function based on the intent
        if intent == "Sqrt":
            math_function = self._kernel.skills.get_function("MathPlugin", "Sqrt")
        elif intent == "Multiply":
            math_function = self._kernel.skills.get_function("MathPlugin", "Multiply")
        else:
            return "I'm sorry, I don't understand."

        # Run the functions in a pipeline
        output = await self._kernel.run_async(
            get_numbers,
            extract_numbers_from_json,
            math_function,
            input_str=request,
        )

        # Create a new context object with the original request
        pipelineVariables = ContextVariables()
        pipelineVariables["original_request"] = request
        pipelineVariables["input"] = request

        # Run the functions in a pipeline with create_response
        output = await self._kernel.run_async(
            get_numbers,
            extract_numbers_from_json,
            math_function,
            create_response,
            input_vars=pipelineVariables,
        )

        return output["input"]

    @sk_function(
        description="Extracts numbers from JSON",
        name="ExtractNumbersFromJson",
    )
    def extract_numbers_from_json(self, context: SKContext):
        numbers = json.loads(context["input"])

        # Loop through numbers and add them to the context
        for key, value in numbers.items():
            if key == "number1":
                # Add the first number to the input variable
                context["input"] = str(value)
            else:
                # Add the rest of the numbers to the context
                context[key] = str(value)

        return context
